keps/sig-etcd/4647-cluster-feature-gate/README.md

```diff
@@ -0,0 +1,468 @@
+
+1. When an etcd server starts, the values of all cluster level features in its `ServerFeatureGate` would saved in the [`member.Attributes`](https://github.com/etcd-io/etcd/blob/e37a67e40b3f5ff8ef81f9de6e7f475f17fda32b/server/etcdserver/api/membership/member.go#L38) as the field `Attributes.feature_gates`. (see the [discussion](#push-vs-poll-when-leader-decides-cluster-feature-gate) about why we choose to push the information through raft)
+
+1. The attributes would then be [published through raft](https://github.com/etcd-io/etcd/blob/e37a67e40b3f5ff8ef81f9de6e7f475f17fda32b/server/etcdserver/server.go#L1745) and stored in the `members` bucket in the backend of all members. Whenever a new member joins or an existing member restarts, its feature gate attributes would be automatically updated in the start up process.
+
+1. After a leader is elected, the leader will decide the values for the cluster level features based on the cluster version and server feature values of all members it receives.
+
+1. The leader sends the final feature values in `ClusterFeatureGateSetRequest` through raft. 
+
+1. Each member apply the updates to their cluster feature gate, and save the results in the `cluster` bucket in the backend.
+
+#### New Raft Proto Changes
+
+```proto
+message Feature {
+  string name = 1;
+  bool enabled = 2;
+}
+
+message Attributes {
+  option (versionpb.etcd_version_msg) = "3.5";
+
+  string name = 1;
+  repeated string client_urls = 2;
+  // the values of all cluster level feature gates set by the configuration of the member server.
+  Feature proposed_feature_gates = 3;
+}
+
+message ClusterFeatureGateSetRequest {
+  option (versionpb.etcd_version_msg) = "3.5";
+
+  // the values of all cluster level feature gates for the whole cluster.
+  Feature feature_gates = 1;
+}
+```
+
+#### How to Reach Consensus of Feature Gates for a New Cluster
+
+![setting cluster feature gate for a new cluster](./cluster_feature_gate_new.png "setting cluster feature gate for a new cluster")
+
+We will leverage the consensus `ClusterVersion` to negotiate the cluster level feature gate values:
+1. if `ClusterVersion` is not set, the cluster feature gate would be `nil`, and all features would be considered disabled. 
+1. when the `ClusterVersion` is set or updated, initialize the cluster feature gate with the `ClusterVersion`. At this point,
+    * if the `member.Attributes` of any members has not been set, use the default values for feature gate at version `ClusterVersion`.
+    * if the `member.Attributes` of all members have been set, take the common set of all the mbmer `--feature-gates` flags and set values of the cluster feature gate: discard any features not recognized at the `ClusterVersion`, and set the feature to false if there is any member setting it to false.
```

Comment:
Don't think we should depend on ClusterVersion here. Just read what member propose.